src/algorithms/StatePreparation.cpp

```diff
@@ -0,0 +1,285 @@
+  double theta{0};
+  double phi{0};
+  double finalT{0};
+  double const magA = std::abs(complexA);
+  double const magB = std::abs(complexB);
+  double const finalR = sqrt(pow(magA, 2) + pow(magB, 2));
+  if (finalR > EPS) {
+    theta = 2 * acos(magA / finalR);
+    double const aAngle = std::arg(complexA);
+    double const bAngle = std::arg(complexB);
+    finalT = aAngle + bAngle;
+    phi = bAngle - aAngle;
+  }
+  return {finalR * exp(std::complex<double>{0, 1} * finalT / 2.), theta, phi};
+}
+
+// works out Ry and Rz rotation angles used to disentangle LSB qubit
+// rotations make up block diagonal matrix U
+[[nodiscard]] auto
+rotationsToDisentangle(std::vector<std::complex<double>> amplitudes)
+    -> std::tuple<std::vector<std::complex<double>>, std::vector<double>,
+                  std::vector<double>> {
+  std::vector<std::complex<double>> remainingVector;
+  std::vector<double> thetas;
+  std::vector<double> phis;
+  for (size_t i = 0; i < (amplitudes.size() / 2); ++i) {
+    auto [remains, theta, phi] =
+        blochAngles(amplitudes[2 * i], amplitudes[2 * i + 1]);
+    remainingVector.push_back(remains);
+    // minus sign because we move it to zero
+    thetas.push_back(-theta);
+    phis.push_back(-phi);
+  }
+  return {remainingVector, thetas, phis};
+}
+
+// creates circuit that takes desired vector to zero
+[[nodiscard]] auto
+gatesToUncompute(std::vector<std::complex<double>> amplitudes, size_t numQubits)
+    -> QuantumComputation {
+  QuantumComputation disentangler{numQubits};
+  for (size_t i = 0; i < numQubits; ++i) {
+    // rotations to disentangle LSB
+    auto [remainingParams, thetas, phis] = rotationsToDisentangle(amplitudes);
+    amplitudes = remainingParams;
+    // perform required rotations
+    bool addLastCnot = true;
+    double const phisNorm = twoNorm(phis);
+    double const thetasNorm = twoNorm(thetas);
+    if (phisNorm != 0 && thetasNorm != 0) {
+      addLastCnot = false;
+    }
+    if (phisNorm != 0) {
+      // call multiplex with RZGate
+      QuantumComputation rzMultiplexer =
+          multiplex(OpType{RZ}, phis, addLastCnot);
+      // append rzMultiplexer to disentangler, but it should only attach on
+      // qubits i-numQubits, thus "i" is added to the local qubit indices
+      for (auto& op : rzMultiplexer) {
+        for (auto& target : op->getTargets()) {
+          target += static_cast<unsigned int>(i);
+        }
+        for (auto control : op->getControls()) {
+          // there were some errors when accessing the qubit directly and
+          // adding to it
+          op->setControls(
+              Controls{Control{control.qubit + static_cast<unsigned int>(i)}});
+        }
+      }
+      disentangler.emplace_back<Operation>(rzMultiplexer.asOperation());
+    }
+    if (thetasNorm != 0) {
+      // call multiplex with RYGate
+      QuantumComputation ryMultiplexer =
+          multiplex(OpType{RY}, thetas, addLastCnot);
```

Comment:
```suggestion
          multiplex(RY, thetas, addLastCnot);
```
simpler and less code to type. applies throughout the whole file.